src/DotNetWorker/Hosting/BootstrapHostBuilder.cs

```diff
@@ -0,0 +1,170 @@
+        private IServiceProviderFactory<object>? _serviceProviderFactory;
+        private bool _defaultsComplete = false;
+
+        public BootstrapHostBuilder(HostApplicationBuilder builder)
+        {
+            _builder = builder;
+
+            foreach (var descriptor in _builder.Services)
+            {
+                if (descriptor.ServiceType == typeof(HostBuilderContext))
+                {
+                    Context = (HostBuilderContext)descriptor.ImplementationInstance!;
+                    break;
+                }
+            }
+
+            if (Context is null)
+            {
+                throw new InvalidOperationException($"{nameof(HostBuilderContext)} must exist in the {nameof(IServiceCollection)}");
+            }
+        }
+
+        public IDictionary<object, object> Properties => Context.Properties;
+
+        public HostBuilderContext Context { get; }
+
+        public IHostBuilder ConfigureAppConfiguration(Action<HostBuilderContext, IConfigurationBuilder> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(Context, _builder.Configuration);
+            }
+            else
+            {
+                _configureAppActions.Add(configureDelegate);
+            }
+
+            return this;
+        }
+
+        public IHostBuilder ConfigureHostConfiguration(Action<IConfigurationBuilder> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(_builder.Configuration);
+            }
+            else
+            {
+                _configureHostActions.Add(configureDelegate);
+            }
+
+            return this;
+        }
+
+        public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(Context, _builder.Services);
+            }
+            else
+            {
+                _configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
```

Comment:
Do we need to have `throw new ArgumentNullException(nameof(configureDelegate)` again when we already check if this is null at the start of the method? It should not be possible for it to be null at this point, no?

Reply:
Nope! Copy/paste error. I've removed it.